File: 973-k-closest-points-to-origin/973-k-closest-points-to-origin.java
class Solution {
    // private int dis(int[] as){
    //     return as[0]*as[0] + as[1]*as[1];
    // }
//     public class Pair implements Comparator<Pair>{
//         int x;
//         int[] ar;
//         Pair(){
            
//         }
//         Pair(int x,int[] ar){
//             this.x=x;
//             this.ar=ar;
//         } 
//         public int compare(Pair a,Pair b){
//             return b.x-a.x;
//         }
//     }
    public int[][] kClosest(int[][] points, int k) {
        PriorityQueue<int[]> pq=new PriorityQueue<int[]>(new Comparator<int[]>(){
            public int compare(int[]p1 ,int[] p2){
                return (p2[0]*p2[0]+p2[1]*p2[1]) - (p1[0]*p1[0]+p1[1]*p1[1]);
            }
        });
        // PriorityQueue<Pair> pq=new PriorityQueue<>(new Pair());

        for(int[] as:points){
            // int x=as[0]*as[0] + as[1]*as[1];
            // Pair rem=new Pair(x,as);
            // pq.add(rem
            pq.add(as);
            if(pq.size()>k) pq.poll();
        }
      int[][] res=new int[k][2];
        int i=0;
       while(pq.size()>0){
          // Pair r=pq.remove();
           res[i]=pq.remove();
           i++;
       }
     return res;
    }
}

File: 973-k-closest-points-to-origin/973-k-closest-points-to-origin.cpp
class Solution {
public:
    vector<vector<int>> kClosest(vector<vector<int>>& points, int k) {
             priority_queue<pair<int,vector<int>>> pq;
      

        for(vector<int> it:points){
//             auto works like we want what type of data type it automitacally determine the data type and give us
            int x= (it[0] * it[0]) + (it[1] * it[1]);
            pq.push({x,it});
            if(pq.size()>k) pq.pop();
        }
      vector<vector<int>> res;
        
       while(pq.size()>0){
       
           auto t=pq.top();
           pq.pop();
          res.push_back(t.second);
          
           
       }
     return res;
    }
};